subprojects/tooling-api/src/testFixtures/groovy/org/gradle/integtests/tooling/fixture/CrossVersionTestEngine.java

```diff
@@ -0,0 +1,234 @@
+        this.id = rootId.append("variant", variant);
+        this.discoveryRequest = request;
+    }
+
+    TestDescriptor discover(final TestEngine discoveryEngine) {
+        return SystemProperties.getInstance().withSystemProperties(systemProperties, new Factory<TestDescriptor>() {
+            @Override
+            public TestDescriptor create() {
+                return discoveryEngine.discover(discoveryRequest, id);
+            }
+        });
+    }
+}
+
+class DelegatingDiscoveryRequest implements EngineDiscoveryRequest {
+    private final EngineDiscoveryRequest delegate;
+    private final List<ClassSelector> selectors = new ArrayList<ClassSelector>();
+
+    DelegatingDiscoveryRequest(EngineDiscoveryRequest delegate) {
+        this.delegate = delegate;
+    }
+
+    void addSelector(ClassSelector selector) {
+        selectors.add(selector);
+    }
+
+    @Override
+    public <T extends DiscoverySelector> List<T> getSelectorsByType(Class<T> selectorType) {
+        if (!selectorType.isAssignableFrom(ClassSelector.class)) {
+            return Collections.emptyList();
+        }
+        return Cast.uncheckedCast(selectors);
+    }
+
+    @Override
+    public <T extends DiscoveryFilter<?>> List<T> getFiltersByType(Class<T> filterType) {
+        return delegate.getFiltersByType(filterType);
+    }
+
+    @Override
+    public ConfigurationParameters getConfigurationParameters() {
+        return delegate.getConfigurationParameters();
+    }
+
+    public EngineDiscoveryListener getDiscoveryListener() {
+        return delegate.getDiscoveryListener();
+    }
+}
+
+class ToolingApiClassloaderDiscoveryRequest extends DelegatingDiscoveryRequest {
+
+    private static final GradleVersion MIN_LOADABLE_TAPI_VERSION = GradleVersion.version("2.6");
+
+    private ToolingApiDistribution toolingApi;
+
+    ToolingApiClassloaderDiscoveryRequest(EngineDiscoveryRequest delegate) {
+        super(delegate);
+        String toolingApiVersionToLoad = getToolingApiVersionToLoad();
+        if (toolingApiVersionToLoad == null) {
+            return;
+        }
+
+        // This is necessary because for the Tooling Api cross version tests, NativeServices
+        // can get initialized in a different classloader, which then makes it broken and unusable
+        // in the test class (because the native shared library is loaded from another classloader).
+        // By initializing it here, we ensure that it is loaded from the classloader the test engine
+        // also uses.
+        NativeServicesTestFixture.initialize();
+
+        for (ClassSelector selector : delegate.getSelectorsByType(ClassSelector.class)) {
+            if (ToolingApiSpecification.class.isAssignableFrom(selector.getJavaClass())) {
+                ClassLoader classLoader = ToolingApiClassLoaderProvider.getToolingApiClassLoader(getToolingApi(toolingApiVersionToLoad), selector.getJavaClass());
```

Comment:
I think the reason we have to reload classes is because Gradle does so here:
https://github.com/gradle/gradle/blob/9b19add723ddfaae7722ee11134955e4517564a4/subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestClassProcessor.java#L89

If we changed that to create `ClassSelectors` from `Strings`, we would only have to replace `Thread.currentThread().getContextClassLoader()` before calling discover/execute.

Reply:
That's right. I initially wanted to change this behavior in Gradle for exactly this use case: https://github.com/gradle/gradle/pull/15786 - perhaps it still makes sense to make this change.

Reply:
I think it would simplify things for us here.

Reply:
As you pointed out below that wouldn't solve the problem. Could you please create a new issue describing your use case for the junit5 repo?

Reply:
https://github.com/junit-team/junit5/issues/2538